jscomp/bsb/bsb_build_util.ml

```diff
@@ -130,7 +130,7 @@ let ( |? ) m (key, cb) = m |> Ext_json.test key cb
 
 type top = Expect_none | Expect_name of string
 
-type package_context = { proj_dir : string; top : top }
+type package_context = { proj_dir : string; top : top; is_pinned: bool }
```

Comment:
Transitive things are stored in `dep_payload`, e.g. both `package_specs` and `jsx` mode have some transitive behaviour.
Not sure it's necessary to put this information in package context.

Reply:
A perhaps more to the point comment: pinned dependencies inherit package specs from the root project, so a bool is not enough. Assuming that's all that's tracked.

Reply:
The way that Rescript handles pinned dependencies is pretty weird. In the existing code, ```pinned-dependencies``` are just a set of strings. After finding all dependencies using ```Bsb_build_util.walk_all_deps```, the compiler iterates over the queue. All dependencies whose name is in that set is passed as a ```bsb_package_kind.Pinned_dependency``` to ```bsb_ninja_regen.regenerate_ninja```. All other dependencies are passed as ```bsb_package_kind.Dependency```. Regardless of whether a dependency is passed as a ```Pinned_dependency``` or a ```Dependency```, it is passed with the root project's ```package_specs``` and ```jsx```. Any entry in ```pinned-dependencies``` which is not the name of a dependency is ignored. 

My PR just generalizes the pre-existing behavior. Firstly I've modified ```Bsb_build_util.walk_all_deps``` so that it correctly traverses the new dependency graph. Secondly, since ```bsb_ninja_regen.regenerate_ninja``` needs to know whether a dependency is pinned, and I already have to calculate that in ```Bsb_build_util.walk_all_deps```, I've modified ```Bsb_build_util.walk_all_deps``` to return that information as well.

To address your comments specifically:
*  While it's unnecessary to put ```is_pinned``` in ```package_context```, I chose to put it there because it makes the code simpler and reduces file reads.
* A bool is enough because it isn't all that's tracked because pinned dependencies are weird.